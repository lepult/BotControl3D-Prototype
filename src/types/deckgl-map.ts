// region Imports
import { Position, Color, TransitionInterpolator } from '@deck.gl/core/typed';
import { IconLayer, PathLayer, PolygonLayer } from '@deck.gl/layers/typed';
import { ScenegraphLayer } from '@deck.gl/mesh-layers/typed';
import { IMergedMapElement, IZone } from './pudu-api/robotMap';
import { CustomDestinationType } from './api/destination';
// endregion

export interface IIconData extends IMergedMapElement {
    name: string;
    position: Position;
    color?: Color;
    angle?: number;
    selected?: boolean;
    routeData: {
        isRouteDestination: boolean;
        isNextDestination: boolean;
        isEarlierDestination: boolean;
        isFinalDestination: boolean;
        isPreviousDestination: boolean;
    },
    customType?: CustomDestinationType;
}

export interface IPathData extends IMergedMapElement {
    name: string;
    path: Array<number>;
    color?: Color;
}

export type ISceneGraphData = {
    name: string;
    position: [number, number];
    angle: number;
    color?: Color;
};

export interface IPolygonData extends IZone {
    name: string;
    color?: Color;
}

export type TTooltip = {
    picked?: boolean;
    object?: IIconData & IPathData & ISceneGraphData & IPolygonData;
    x?: number;
    y?: number;
};

export type TLayers = (IconLayer<IIconData, NonNullable<unknown>>
    | PathLayer<IPathData, NonNullable<unknown>>
    | ScenegraphLayer<ISceneGraphData, NonNullable<unknown>>
    | PolygonLayer<IPolygonData, NonNullable<unknown>>)[];

export type TViewState = {
    bearing: number,
    latitude: number,
    longitude: number,
    pitch: number,
    zoom: number,
    maxZoom?: number,
    minZoom?: number,
    rotationX?: number,
    maxPitch?: number,
    minPitch?: number,
    transitionDuration?: number,
    transitionInterpolator?: TransitionInterpolator,
}

export enum MapRobotStatus {
    'Offline' = 'Offline',
    'Idle' = 'Idle',
    'Charged' = 'Charged',
    'Charging' = 'Charging',
    'ChargingError' = 'ChargingError',
    'ArrivedAtDiningOutlet' = 'ArrivedAtDiningOutlet',
    'SendToChargingPile' = 'SendToChargingPile',
    'WaitForDoor' = 'WaitForDoor',
    'WaitForElevator' = 'WaitForElevator',
    'MovingWithElevator' = 'MovingWithElevator',
    'ArrivedAtPickupDestination' = 'ArrivedAtPickupDestination',
    'PrepareDriveToDestination' = 'PrepareDriveToDestination',
    'SendToDestination' = 'SendToDestination',
    'ArrivedAtDestination' = 'ArrivedAtDestination',
    'Avoid' = 'Avoid',
    'Blocked' = 'Blocked',
    'Cancel' = 'Cancel',
    'Error' = 'Error',
    'Pause' = 'Pause',
}

export enum PreviewType {
    Robot,
    Floor,
}

export enum DragMode {
    translate = 'translate',
    rotate = 'rotate',
}

export type TUndoStackItem = {
    id: string,
    position: [number, number, number],
    orientation: [number, number, number],
}

export type TRobotLayerData = {
    name: string,
    robotId: string,
    position: [number, number],
    icon: string,
    color: Color,
    orientation: [number, number, number],
};
