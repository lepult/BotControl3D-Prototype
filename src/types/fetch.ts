export enum FetchState {
    initial,
    pending,
    fulfilled,
    rejected,
}
