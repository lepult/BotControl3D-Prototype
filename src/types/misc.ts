export enum UserType {
    guest,
    team,
    admin,
}
