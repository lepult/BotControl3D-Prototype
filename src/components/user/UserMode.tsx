/* eslint-disable @typescript-eslint/ban-ts-comment */
import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { setRefreshScrollEnabled } from 'chayns-api';
// @ts-ignore
import { SmallWaitCursor } from 'chayns-components';
import { ChaynsViewMode, removeFooter, updateChaynsViewmode } from '../../utils/pageSizeHelper';
import UserModeMap from './user-mode-map/UserModeMap';
import UserModeButtons from './user-mode-buttons/UserModeButtons';
import { selectMapsFetchState, selectSelectedMap } from '../../redux-modules/map/selectors';
import { FetchState } from '../../types/fetch';

const UserMode = () => {
    useEffect(() => {
        updateChaynsViewmode(ChaynsViewMode.wide);
        removeFooter(true);
        void setRefreshScrollEnabled(false);
    }, []);

    const selectedMap = useSelector(selectSelectedMap);
    const mapFetchState = useSelector(selectMapsFetchState);

    return (
        <div>
            <UserModeButtons/>
            {mapFetchState === FetchState.pending && (
                <div
                    style={{
                        display: 'flex',
                        justifyContent: 'center',
                        alignItems: 'center',
                        height: '100%',
                        width: '100%'
                    }}
                >
                    <SmallWaitCursor show/>
                </div>
            )}
            {selectedMap
                ? <UserModeMap mapId={selectedMap}/>
                : <div>Fehler</div>}

        </div>
    );
};

export default UserMode;
