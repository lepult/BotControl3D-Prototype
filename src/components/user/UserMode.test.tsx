/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react';
import renderer from 'react-test-renderer';
import UserMode from './UserMode';
import { FetchState } from '../../types/fetch';

const reduxState = {
    map: {
        selectedMap: 1,
        fetchState: FetchState.fulfilled,
    }
}

jest.mock('react-redux', () => ({
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-return
    useSelector: (selector) => selector(reduxState),
}));

jest.mock('../map/user-mode-buttons/UserModeButtons', () => () => <div className="buttons"/>);
jest.mock('chayns-components', () => ({
    SmallWaitCursor: () => <div className="wait-cursor"/>,
}));
jest.mock('../map/Map', () => () => <div className="map"/>);

it('selected map should result in displayed map', () => {
    const component = renderer.create(
        <UserMode/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
it('fetching maps should result in wait cursor and no map', () => {
    reduxState.map.fetchState = FetchState.pending;
    reduxState.map.selectedMap = 0;

    const component = renderer.create(
        <UserMode/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    reduxState.map.fetchState = FetchState.fulfilled;
});
