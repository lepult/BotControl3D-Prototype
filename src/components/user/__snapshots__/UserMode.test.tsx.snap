// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetching maps should result in wait cursor and no map 1`] = `
<div>
  <div
    className="buttons"
  />
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": "100vh",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <div
      className="wait-cursor"
    />
  </div>
  0
</div>
`;

exports[`selected map should result in displayed map 1`] = `
<div>
  <div
    className="buttons"
  />
  <div
    className="map"
  />
</div>
`;
